File: src/competitve practice/leetcode-30day-challenge/week4/JumpGame.java
class Solution {
    public boolean canJump(int[] nums) {
        if (nums[0] == 0)
            if (nums.length == 1)
                return true;
            else
                return false;

        int maxVal = nums[0];
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != 0) {
                if (maxVal <= nums[i]) {
                    maxVal = nums[i];
                    continue;
                }
                maxVal--;
                continue;
            } else {
                if (i == nums.length - 1)
                    continue;

                if (maxVal > 0) {
                    maxVal--;
                    if (maxVal == 0)
                        return false;
                    continue;
                }

                return false;
            }
        }
        return true;
    }
}

File: LeetCode/leetcode-30day-challenge/week4/C++/JumpGame.cpp
#include <bits/stdc++.h>
using namespace std;

//Type: Math
//Approach: Start from the 0th index of the array
//          Maintain the maximum reachable distance from that cell at every index
//          If at any index, the maximum reachable distance is less than equal to current index,
//          we can not reach the end. Do handle the corner cases -> start and the last ind :)
//Complexity: O(N) time

class Solution
{
public:
    bool canJump(vector<int> &nums)
    {
        int maxReachable = 0;
        if (nums.size() == 1)
        {
            return 1;
        }
        for (int i = 0; i < nums.size(); i++)
        {
            maxReachable = max(maxReachable, i + nums[i]);
            if (i == nums.size() - 1)
            {
                return true;
            }
            if (maxReachable <= i)
            {
                return false;
            }
        }
        return true;
    }
};